### PR TITLE
add support to eval code and solve file:// url bug

### PR DESCRIPTION
**Pull request by Alphapage**

### CODE REVIEW

**Comment by Alphapage** [comment]
This PR could solve: 
https://github.com/Microsoft/vscode-chrome-debug-core/issues/205
https://github.com/Microsoft/vscode-node-debug2/issues/122
maybe https://github.com/Microsoft/vscode-chrome-debug/issues/460 and older issues I found.

**Comment by Alphapage** [comment]
For info, I use global to register evaled sources. This is ugly, but modules were not working. Maybe different process !

**Comment by Alphapage** [comment]
The main problems I was unable to solve:
- you need to  add `debugger;` at the beginning of eval code or add a breakpoint in a non eval code otherwise breakpoints will never be applied.
- it only works with node-debug2. node-debug or chrome-debug shows the evaled code instead of mapped one
- you need to show a link somewhere in console or in UI to be able to open the evaled code in a tab. I wasn't able to find a way to do it.

File: src/chrome/chromeDebugAdapter.ts [hunk 1]
```diff
@@ -1423,6 +1423,10 @@ export abstract class ChromeDebugAdapter implements IDebugAdapter {
             // If the frame doesn't have a function name, it's either an anonymous function
             // or eval script. If its source has a name, it's probably an anonymous function.
             const frameName = functionName || (script.url ? '(anonymous function)' : '(eval code)');
+            if (script.hasSourceURL) {
+                if (!global["evalSources"]) global["evalSources"] = [];
```
Comment (roblourens) [inline src/chrome/chromeDebugAdapter.ts]: Please don't use `global`, try to find another way to share this list between modules.

File: src/sourceMaps/sourceMap.ts [hunk 2]
```diff
@@ -161,6 +161,7 @@ export class SourceMap {
      * Will return null instead of a mapping on the next line (different from generatedPositionFor).
      */
     public authoredPositionFor(line: number, column: number): MappedPosition {
+        if (global['evalSources'] && global['evalSources'].find(s => this._generatedPath == s) && line > 0) line--; // check if eval code
```
Comment (roblourens) [inline src/sourceMaps/sourceMap.ts]: Why do you need to adjust the line differently based on whether it's an eval script or not?

File: src/sourceMaps/sourceMap.ts [hunk 3]
```diff
@@ -223,7 +224,7 @@ export class SourceMap {
             return null;
         } else {
             return {
-                line: position.line - 1, // Back to 0-indexed lines
+                line: global["evalSources"] && global["evalSources"].find(s => s == this._generatedPath) ? position.line : position.line - 1, // check if eval code else Back to 0-indexed lines
```
Comment (roblourens) [inline src/sourceMaps/sourceMap.ts]: Same here - the 0-indexed or 1-indexed lines are a property of the source-map lib, it doesn't matter whether the code originated as an eval script or not.

**Comment by roblourens** [comment]
Please see tslint issues as well

File: src/sourceMaps/sourceMap.ts (hunk 2 above)
Comment (Alphapage) [inline src/sourceMaps/sourceMap.ts]: I think when code is an eval script, then `(function(){\n` is added and chrome debugger result is offset by 1 line vs source-map which use the script code.

File: src/chrome/chromeDebugAdapter.ts (hunk 1 above)
Comment (Alphapage) [inline src/chrome/chromeDebugAdapter.ts]: I tried import/export without success. Do you have another idea ?

**Comment by Alphapage** [comment]
I don't know how chrome debugger works, but the best way to solve this problem would be to send the eval file from //#sourceURL and sourcemap file from //#sourceMappingURL to chrome debugger and let chrome debugger do the job, apply breakpoints...
But I don't know if v8 debugger api provides such a feature.

File: src/sourceMaps/sourceMap.ts (hunk 3 above)
Comment (Alphapage) [inline src/sourceMaps/sourceMap.ts]: Same as previous answer. I think either generated and authored code relay on`(function(){\n` problem.
in my opinion, it is more +1 indexed than 1-indexed: I keep the base index but add some adjustments.

**Pull request closed without merging**